pytensor/tensor/basic.py

```diff
@@ -2758,29 +2758,21 @@ def concatenate(tensor_list, axis=0):
     return join(axis, *tensor_list)
 
 
-def horizontal_stack(*args):
+def hstack(*args):
     r"""Stack arrays in sequence horizontally (column wise)."""
-    # Note: 'horizontal_stack' and 'vertical_stack' do not behave exactly like
-    # Numpy's hstack and vstack functions. This is intended, because Numpy's
-    # functions have potentially confusing/incoherent behavior (try them on 1D
-    # arrays). If this is fixed in a future version of Numpy, it may be worth
-    # trying to get closer to Numpy's way of doing things. In the meantime,
-    # better keep different names to emphasize the implementation divergences.
 
     if len(args) < 2:
```

Comment:
Numpy also doesn't raise on a single input. As the (deleted) comment says, it has somewhat unexpected behaviors:

- `np.hstack([flat_array]) == flat_array`
- `np.vstack([flat_array]) == flat_array[None]`
- `np.dstack([flat_array]) == flat_array[:, None]`

If we're going to match the names I think it's important we dig into the internal logic of these three functions a little bit and try to replicate it. I don't think anyone's code depends on these behaviors, but they are idiosyncrasies of how they are doing the insert_dims and concatenate `Op`.

Reply:
@jessegrabowski 
np.dstack uses np.atleast_3d to make sure the array is at least 3-dimensional
However, np.atleast_3d behaves differently than pt.atleast_3d, see below code for reference

```
a = np.array([1, 2, 3, 4])

print(a.shape) # (4,)
print(pt.atleast_3d(a).eval().shape) # (1, 1, 4)
print(np.atleast_3d([a]).shape) # (1, 4, 1)
```

Should pt.atleast_3d be changed to behave like np.atleast_3d?

Reply:
Yeah I have come across this behavior and it's annoying. We should behave like numpy but we need a transition strategy.

I suggest we had a kwarg "expand_left” that for now is `None` and behaves like before but with a warning the default will be False and users have to specify it explicit.

```python
det atleast_3d(x, expand_left: Optional[bool] = None):
  if expand_left is None:
    expand_left = True
    if x.type.ndim == 2:
      warnings.warning("The behavior of at_least3d will change for 2d arrays to match that of numpy. If you want to keep the old behavior specify expand_left=True", FutureWarning)
  
  # Act based on expand_left flag
  ...
```

Reply:
pt.atleast_3d uses pt.atleast_Nd
pt.atleast_Nd already has an arg `left` which is true by default.

An example:

```
a = np.array([1, 2, 3, 4])

print(pt.atleast_3d([a], left=True).eval().shape) # (1, 1, 4)
print(pt.atleast_3d([a], left=False).eval().shape) # (1, 4, 1)
print(np.atleast_3d([a]).shape) # (1, 4, 1)
```

To match numpy behaviour the array must be inside a list

Reply:
Nice so we should add the same warning around the left key

Reply:
I will use this property only for pt.dstack
What warning should be added?